### ISSUE TITLE
Solve stumbles over icu constraints

### ISSUE THREAD

**Issue by h-vetinari** [conda-forge/libxml2-feedstock#137]
In https://github.com/conda-forge/libpdal-feedstock/pull/17, the build started failing on windows after #121; after #136, it now fails on linux-64 and linux-ppc64le (though not aarch64; I have the suspicion that some solver idiosyncracies w.r.t. build-string ordering might be the reason).

The problem is that `libpdal` depends transitively on `icu` (via `libgdal-core` -> `xerces-c`), and while the build of `libpdal` passes, several outputs that depend on `{{ pin_subpackage('libpdal-core', exact=True) }}` as well as something that wants libxml2 then fail with ([logs](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1068467&view=logs&j=f4da4d1f-390b-5521-e6df-2f06cb62cc8d&t=caf1a362-b66b-5153-f05a-912b513ef6fc))

```
Packaging libpdal-arrow
[...]
Could not solve for environment specs
The following packages are incompatible
├─ libpdal-core 2.8.1.* h8f62525_1 is installable and it requires
│  └─ libgdal-core >=3.9.3,<3.10.0a0 , which requires
│     └─ xerces-c >=3.2.5,<3.3.0a0  with the potential options
│        ├─ xerces-c 3.2.5 would require
│        │  └─ icu >=75.1,<76.0a0 , which can be installed;
│        └─ xerces-c 3.2.5 would require
│           └─ icu >=73.2,<74.0a0 , which can be installed;
└─ libxml2 2.13.4.* h064dc61_2 is not installable because it requires
   └─ icu <0.0a0 , which conflicts with any installable versions previously reported.
```

It's quite clear that a solution (`libxml2-2.13.4-hb346dea_2`) exist, and the fact that the solver insists on `h064dc61_2` (that forces `icu<0.0`) is likely a solver bug. Still, it seems to be the current state of the world... 😑 

Not even adding an `icu` dependency to `libpdal` worked, but still ran into ([logs](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1069968&view=results))
```
Could not solve for environment specs
The following packages are incompatible
├─ libpdal-core 2.8.1.* h4ea88c3_1 is installable and it requires
│  └─ icu >=75.1,<76.0a0 , which can be installed;
└─ libxml2 2.13.4.* h064dc61_2 is not installable because it requires
   └─ icu <0.0a0 , which conflicts with any installable versions previously reported.
```

Working iteratively through the failing outputs, what finally ended up solving things was the following diff

<details>

```diff
--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -30,6 +30,11 @@ requirements:
     - zlib
     - libxml2
     - proj
+  run:
+    # help solver choose icu-comptible libxml2, because libgdal-core
+    # needs xerces-c which depends on icu on unix, c.f. also
+    # https://github.com/conda-forge/libxml2-feedstock/pull/121
+    - icu *     # [unix]
   run_constrained:
     - pdal {{ version }}.*

@@ -272,6 +277,9 @@ outputs:
         - PDAL_PLUGIN_NAME=tiledb
       run_exports:
         - {{ pin_subpackage("libpdal-tiledb", max_pin="x.x") }}
+      ignore_run_exports_from:
+        # hide icu from being picked up by smithy
+        - {{ "icu" }}
     requirements:
       build:
         - {{ compiler('cxx') }}
@@ -281,6 +289,8 @@ outputs:
         - ninja
       host:
         - tiledb
+        # see comment w.r.t. libxml2 & icu at the top
+        - icu *             # [unix]
         - {{ pin_subpackage('libpdal-core', exact=True) }}
       run:
         - {{ pin_subpackage('libpdal-core', exact=True) }}
@@ -345,6 +355,9 @@ outputs:
         - PDAL_PLUGIN_NAME=arrow
       run_exports:
         - {{ pin_subpackage("libpdal-arrow", max_pin="x.x") }}
+      ignore_run_exports_from:
+        # hide icu from being picked up by smithy
+        - {{ "icu" }}
     requirements:
       build:
         - {{ compiler('cxx') }}
@@ -356,6 +369,8 @@ outputs:
         - libarrow
         - libarrow-dataset
         - libparquet
+        # see comment w.r.t. libxml2 & icu at the top
+        - icu *             # [unix]
         - {{ pin_subpackage('libpdal-core', exact=True) }}
       run:
         - {{ pin_subpackage('libpdal-core', exact=True) }}
@@ -423,6 +438,9 @@ outputs:
       skip: true  # [win]
       run_exports:
         - {{ pin_subpackage("libpdal-cpd", max_pin="x.x") }}
+      ignore_run_exports_from:
+        # hide icu from being picked up by smithy
+        - {{ "icu" }}
     requirements:
       build:
         - {{ compiler('cxx') }}
@@ -434,6 +452,8 @@ outputs:
       host:
         - cpd
         - fgt
+        # see comment w.r.t. libxml2 & icu at the top
+        - icu *             # [unix]
         - {{ pin_subpackage('libpdal-core', exact=True) }}
       run:
         # TODO: add run_exports to cpd
```

</details>

I'm worried though that other downstream users will run into similar solver issues when trying to combine something that ends up pulling in the no-icu `libxml2` as well as a package that depends on `icu`. 

CC @ocefpaf @isuruf

**Comment by isuruf** [comment]
This is an issue with how conda-build uses the solver. For environments, it first solves the environment without the pinned subpackage and then tries to install the subpackage with the environment fully solved which leads to this error. This is IMO a bug in conda-build.

**Comment by h-vetinari** [comment]
> This is IMO a bug in conda-build.

Could you raise an issue?

**Comment by rouault** [comment]
@isuruf I was considering opening a new issue against conda-build but likely means the technical baggage to express it properly. While researching for exiting issues matching "pinned subpackage", I found  https://github.com/conda/conda-build/issues/3308 which might be related

**Comment by isuruf** [comment]
Closing as a duplicate of the above conda-build PR
